Add an espeak+video backend

197g opened the pull request:
Depends on external programs `ffmpeg`, `sox` and `espeak-ng`. Example in `examples/mp4.md`.

197g committed:
Introduce new backend specializing beamer

This required significantly less work than imagined. Also, it prepares
open and close calls for customization as we'll want to be notified of
every opened frame to keep track of the index for synchronizing audio to
the frames which we will later extract individually from the built pdf.
This type parameter _somehow_ defaults to `()` when it is not provided
which means there is no need to change _any_ other site of use by simply
adding the respective impls for the unit tuple as well.

197g committed:
Wire up the high-level logic

197g committed:
Complete the full output generation flow

197g committed:
Bug fixes, bug fixes, bug fixes

197g committed:
Switch to Henrique, because I liked that voice

197g committed:
Fix off-by-one in frame numbering

197g committed:
Add an example utilizing video output

oberien commented on src/backend/ffmpeg/mod.rs:
```diff
@@ -0,0 +1,241 @@
+}
+
+impl SlidesFfmpegEspeak {
+    fn open_speech_file(&self, cfg: &Config, diagnostics: &Diagnostics<'_>) -> Result<File> {
```
```suggestion
    fn create_speech_file(&self, cfg: &Config, diagnostics: &Diagnostics<'_>) -> Result<File> {
```

oberien commented on src/backend/ffmpeg/mod.rs:
```diff
@@ -0,0 +1,241 @@
+            Ok(file) => Ok(file),
+            Err(e) => {
+                diagnostics
+                    .error("error creating temporary espeak file for frame")
```
While for now it is creating a *temporary* espeak file, it most likely won't be in the long run. Instead, it'll just be in the out_dir as everything else should. Would it be possible to rephrase the error message accordingly? Otherwise we can revisit it when actually implementing proper caching, out_dir handling etc.

HeroicKatora commented:
Despite having the right timing to be an April's fool joke, the suggestion is somewhat serious.

oberien commented:
I really like this PR!!! Even if it was just for April Fool's, I'd have merged it. I'll review it soon-ish™.

Aren't there any TTS engines, which don't sound this robotic and annoying? For example I've seen twitch streamer bots with quite good TTS.

HeroicKatora commented:
There are alternatives, many of them backed by cloud infrastructure which I wanted to avoid. Streamers are using a system provided by streamlabs that's ran on their servers and I can't tell you which voice models it uses. I only found this out when looking it up right now but it explains why you aren't able to customize or prefilter messages. There are of course a few possibilities that are fully open source and run locally (please edit the list if you find any other):

* https://github.com/festvox
* https://github.com/CSTR-Edinburgh/merlin

oberien commented on src/backend/latex/document/beamer.rs:
```diff
@@ -17,58 +17,83 @@ pub struct Beamer {
     /// Closes the beamerboxesrounded / slides until including the given level.
     /// Also performs the according checks and updates the heading stack.
-    pub fn close_until(
+    pub fn close_until<C: Default + Extend<FrameEvent>>(
```
Just return a Vec, we shouldn't need anything further than that (for now without optimizations), and it makes everything simpler and more readable.

197g replied:
With this style, the compiler defaults this parameter to a simple `()` if it is not provided and the method is used as a `;`-terminated statement. This is due to magic. It was convenient though and means no allocation when the result is unused which was very satisfying as it just happens to click together nicely.

oberien replied:
Dead-code analysis should also get rid of the allocations. And even if not, I think for now clarity is more important than performance, especially when we're talking about a few allocations over the whole lifetime of the program. Our bottleneck is not the rust part, but instead latex. Optimizations can come later.

197g replied:
The only kinds of allocations for which I know that llvm's dead code analysis actually can get rid of are `alloca`. Whether it considers others a side effect, or there is some bad interaction going on but it is _very_ hard for it to reason about any other kinds of allocator use. It's annoying.

197g replied:
Actually that is not exactly true. A direct pair of `alloc`/`dealloc` is optimized away but even one call to `realloc` in-between seems to confuse it enough to give up. [See this playground](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=3f8c116624e0a4ffe385d592f8a743e1).

oberien replied:
Yeah, I remember llvm being weird around `malloc` and `free`, most of the times treating them like black boxes unfortunately. I can definitely see that your approach is pretty fancy and cool and deep dark magic and all, and I also really love those solutions, but I think in this case it doesn't add much (for now) while hurting readability. Premature optimization is the root of all evil ;)

oberien commented on src/backend/ffmpeg/mod.rs:
```diff
@@ -0,0 +1,241 @@
+}
+
+impl Extend<BeamerFrameEvent> for CurrentFrame {
+    fn extend<Iter: IntoIterator<Item=BeamerFrameEvent>>(&mut self, iter: Iter) {
```
While `Extend` might fit functionally, I think it doesn't really imply the intent of this action. What do you think about having a method instead, with a name like `advance_with` or something expressing its function better?

197g replied:
The original intent was to use `FromIterator`. I could rewrite `close_until` to produce an iterator that can be used but it is slightly more involved. The next best trait was `Extend` which is somewhat more powerful but, imo, serves a very similar purpose in the sense that it is mostly transparent to the caller which implementation is being used.

oberien replied:
I understood that reasoning, having tried to rewrite it as `FromIterator`, which doesn't work easily. Extend definitely is a nice solution generically, but I still think it hurts ease of understanding the code.

oberien commented on src/backend/ffmpeg/mod.rs:
```diff
@@ -0,0 +1,241 @@
+            }
+        }
+
+        <<Beamer as Backend<'static>>::CodeBlock as CodeGenUnit<'a, CodeBlock<'a>>>
```
1. No need to specify specific lifetimes here, you can replace both `'static` and `'a` with `'_`.
2. The compiler's inference goes even further, you can just write `CodeGenUnit::new` directly here.

197g replied:
Hm, weird. I remember that inference had failed somewhere. But I'll remove those superficial annotations then.

oberien commented on src/backend/ffmpeg/mod.rs:
```diff
@@ -0,0 +1,241 @@
+        match self {
+            CodeBlockGen::Speech(file) => Some(file),
+            CodeBlockGen::Normal(inner) => 
+                <<Beamer as Backend<'static>>::CodeBlock as CodeGenUnit<'a, CodeBlock<'a>>>
```
~

oberien commented on src/backend/ffmpeg/mod.rs:
```diff
@@ -0,0 +1,241 @@
+    ) -> Result<()> {
+        match self {
+            CodeBlockGen::Normal(inner) =>
+                <<Beamer as Backend<'static>>::CodeBlock as CodeGenUnit<'a, CodeBlock<'a>>>
```
~

oberien commented on src/backend/latex/document/beamer.rs:
```diff
@@ -17,58 +17,83 @@ pub struct Beamer {
     headings: Vec<i32>,
 }
 
+pub enum FrameEvent {
```
Compiler warnings say that you should derive `Debug` and `Copy` here.

oberien commented on src/backend/latex/document/beamer.rs:
```diff
@@ -17,58 +17,83 @@ pub struct Beamer {
     /// Opens the beamerboxesrounded / slides until including the given level, updating the heading
     /// stack.
-    pub fn open_until(
+    pub fn open_until<C: Default + Extend<FrameEvent>>(
```
~

oberien commented on src/backend/latex/document/beamer.rs:
```diff
@@ -179,3 +204,9 @@ impl<'a> Backend<'a> for Beamer {
+    fn extend<Iter: IntoIterator<Item=FrameEvent>>(&mut self, iter: Iter) {
+        for _ in iter {}
+    }
+}
```
Not needed if you just use a `Vec`.

oberien commented on src/main.rs:
```diff
@@ -128,6 +148,108 @@ fn gen(cfg: &Config, markdown: String, out: impl Write) {
+    // attempting to do but, I assumed, since they can't admit to the PHP crowds of SaaS image
+    // tools that their tools are utterly messy and completely unmaintainable they introduce a
+    // black-list that makes it impossible to use for other people as well.
+    // /rant
```
:+1:

oberien commented on src/main.rs:
```diff
@@ -128,6 +148,108 @@ fn gen(cfg: &Config, markdown: String, out: impl Write) {
+        .arg(pdf.as_ref())
+        .arg("pages")
+        .status()
+        .expect("Converting pdf with `convert` failed");
```
Considering the rant above, I'm a bit surprised by this error message :thinking: ;)

197g replied:
Don't you see, I'm expecting that converting with `convert` fails :see_no_evil: No, I just missed this message while replacing it.

oberien commented on src/main.rs:
```diff
@@ -128,6 +148,108 @@ fn gen(cfg: &Config, markdown: String, out: impl Write) {
+    // /rant
+    Command::new("pdftoppm")
+        .current_dir(&cfg.out_dir)
+        .arg("-png")
+        .arg(pdf.as_ref())
```
```suggestion
        .args(&["-png", pdf.as_ref()])
```
I think grouping args here eases readability.

197g replied:
This one is intentional. `pdf.as_ref()` is `&Path` while the former is `&str` so we would have to wrap either with more explicit converters which would, imo, negate some of the readability gained from grouping these arguments.

oberien commented on src/main.rs:
```diff
@@ -128,6 +148,108 @@ fn gen(cfg: &Config, markdown: String, out: impl Write) {
+        let wav = format!("espeak-{}.wav", idx);
+
+        if !cfg.out_dir.join(&frame).exists() {
+            break;
```
Shouldn't this be an ICE / output an ICE message?

197g replied:
This was the sloppy implementation for detecting the number of frames :)

oberien replied:
Ohhh, I see. Ewwww!!! At least add a TODO :D

oberien commented on src/main.rs:
```diff
@@ -128,6 +148,108 @@ fn gen(cfg: &Config, markdown: String, out: impl Write) {
+
+    let mut audios = vec![];
+
+    for idx in 0.. {
```
TOCTOU, but should be good enough for our case.

oberien commented on src/main.rs:
```diff
@@ -128,6 +148,108 @@ fn gen(cfg: &Config, markdown: String, out: impl Write) {
+
+        Command::new("espeak-ng")
+            .current_dir(&cfg.out_dir)
+            .arg("-f")
+            .arg(&speak)
+            .arg("-w")
+            .arg(&wav)
```
```suggestion
            .args(&["-f", &speak])
            .args(&["-w", &wav])
```

oberien commented on src/main.rs:
```diff
@@ -128,6 +148,108 @@ fn gen(cfg: &Config, markdown: String, out: impl Write) {
+            .status()
+            .expect("Conversion with `espeak-ng` failed.");
+
+        let output = Command::new("sox")
```
If we depend on `ffmpeg` anyway, why don't we use `ffprobe` / the ffmpeg tool suite to get the wav's metadata?

197g replied:
It was non-obvious how to do this with `ffmpeg`, if you have some pointers to the documentation then I'll gladly switch the implementation.

oberien replied:
While it's possible with `ffmpeg`, using `ffprobe` is probably easier (and `ffprobe` is also part of (at least the arch) ffmpeg installs. You can see both solutions [here](https://superuser.com/a/945604).

oberien commented on src/main.rs:
```diff
@@ -128,6 +148,108 @@ fn gen(cfg: &Config, markdown: String, out: impl Write) {
+
+    // concatenate all audio
+    {
+        let mut concat = Command::new("sox");
```
Again, `ffmpeg` could do this as well.

197g replied:
That's right, and the concat filter is already used as well. That one should be easy to switch.

oberien reviewed (commented):
The slides in the output seem to have a pretty bad quality, could we export them with a better quality / in a higher resolution?

I think the error messages when a binary (dependency) doesn't exist could be improved.

NB: What do you think about adding an `espeak` backend in general, which generates a `flac`, `mp3`, … from just normal markdown?

197g committed:
Apply code clarity feedback

Remove type parameter enabling efficient frame tracking for the sake of
readability from a single return type, allocating vectors of a couple
hundred elements should not be the main cause of slowdown.

Removes unnecessary explicit trait impl syntax where calling methods on
the trait would correctly deduce all parameters and the type itself.

Add missed derives of `Copy`, `Clone`, `Debug`.

197g committed:
Switch from sox to ffmpeg

This just reduces the number of dependencies.

197g committed:
Address UX of missing dependencies

Also addresses the comments on a few suboptimal parts of the
implementation which are expected to change in the future.

oberien commented on src/main.rs:
```diff
@@ -128,6 +150,155 @@ fn gen(cfg: &Config, markdown: String, out: impl Write) {
+    // /rant
+    Command::new("pdftoppm")
+        .current_dir(&cfg.out_dir)
+        .arg("-png")
```
For better Resolution of the images in the mp4 (currently they only have 100 or 150 dpi, which looks pixelated).

```suggestion
        .args(&["-png", "-rx", "600", "-ry", "600"])
```

oberien reviewed (changes requested):
The changes look good to me so far. See comment inline.

oberien committed:
Improve espeak video image quality
